core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -317,250 +340,240 @@ V put(K key, V value, Function<ByteBuffer, V> transformer) {
-            rebalance(c);
-            return put(key, value, transformer);
-        }
+            if (!c.publish()) {
+                c.releaseValue(newValueReference);
+                rebalance(c);
+                continue;
+            }
 
-        if (!finishAfterPublishing(opData, c)) {
-            return put(key, value, transformer);
+            if (c.linkValue(opData) != TRUE) {
+                c.releaseValue(newValueReference);
+                c.unpublish();
+            } else {
+                c.unpublish();
+                checkRebalance(c);
+                return null;
+            }
         }
-
-        return null;
     }
 
     Result<V> putIfAbsent(K key, V value, Function<ByteBuffer, V> transformer) {
         if (key == null || value == null) {
             throw new NullPointerException();
         }
 
-        Chunk<K, V> c = findChunk(key); // find chunk matching key
-        Chunk.LookUp lookUp = c.lookUp(key);
-        if (lookUp != null && lookUp.valueSlice != null) {
-            if (transformer == null) {
-                return Result.withFlag(FAILURE);
-            }
-            Result<V> res = ValueUtils.transform(lookUp.valueSlice, transformer);
-            if (res.operationResult == SUCCESS) {
-                return res;
-            }
-            return putIfAbsent(key, value, transformer);
-        }
+        while (true) {
+            Chunk<K, V> c = findChunk(key); // find chunk matching key
+            Chunk.LookUp lookUp = c.lookUp(key);
 
-        // if chunk is frozen or infant, we can't add to it
-        // we need to help rebalancer first, then proceed
-        Chunk.State state = c.state();
-        if (state == Chunk.State.INFANT) {
-            // the infant is already connected so rebalancer won't add this put
-            rebalance(c.creator());
-            return putIfAbsent(key, value, transformer);
-        }
-        if (state == Chunk.State.FROZEN || state == Chunk.State.RELEASED) {
-            rebalance(c);
-            return putIfAbsent(key, value, transformer);
-        }
+            if (lookUp != null && lookUp.valueSlice != null) {
+                if (updateVersionAfterLinking(c, lookUp)) {
+                    continue;
+                }
+                if (transformer == null) {
+                    return Result.withFlag(FALSE);
+                }
+                Result<V> res = operator.transform(lookUp.valueSlice, transformer, lookUp.version);
+                if (res.operationResult == TRUE) {
+                    return res;
+                }
+                continue;
+            }
 
+            // if chunk is frozen or infant, we can't add to it
+            // we need to help rebalancer first, then proceed
+            if (inTheMiddleOfRebalance(c) || finalizeDeletion(c, lookUp)) {
+                continue;
+            }
+            int oldVersion = lookUp == null ? INVALID_VERSION : -Math.abs(lookUp.version);
```

Comment:
Please add comment explaining why do we take the minus of the absolute value. Why minus indeed?

Reply:
I added comments where they are needed.